File: backend/src/main/java/com/seuprojeto/progress/ProgressService.java
package com.seuprojeto.progress;

import com.seuprojeto.catalog.Content;
import com.seuprojeto.catalog.ContentRepository;
import com.seuprojeto.catalog.ContentService;
import com.seuprojeto.catalog.Episode;
import com.seuprojeto.catalog.EpisodeRepository;
import com.seuprojeto.progress.dto.ProgressResponse;
import com.seuprojeto.user.User;
import com.seuprojeto.user.UserService;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
public class ProgressService {

    private final UserProgressRepository progressRepository;
    private final EpisodeRepository episodeRepository;
    private final ContentRepository contentRepository;
    private final UserService userService;
    private final ContentService contentService;

    @Transactional
    public ProgressResponse marcarEpisodio(Long episodioId, ProgressStatus status, String userEmail) {
        User user = userService.findByEmail(userEmail);
        Episode episode = episodeRepository.findById(episodioId)
                .orElseThrow(() -> new EntityNotFoundException("Episódio não encontrado"));

        UserProgress progress = progressRepository
                .findByUsuarioIdAndEpisodioId(user.getId(), episodioId)
                .orElse(UserProgress.builder()
                        .usuario(user)
                        .episodio(episode)
                        .build());

        progress.setStatus(status);
        progress.setAtualizadoEm(LocalDateTime.now());
        progressRepository.save(progress);

        Long conteudoId = episode.getTemporada().getConteudo().getId();
        Long temporadaId = episode.getTemporada().getId();

        long totalTemporada = episodeRepository.countByTemporadaId(temporadaId);
        long assistidosTemporada = progressRepository.countByUsuarioIdAndStatusAndEpisodio_Temporada_Id(
                user.getId(), ProgressStatus.ASSISTIDO, temporadaId);
        double progressoTemporada = totalTemporada == 0 ? 0.0 : (100.0 * assistidosTemporada / totalTemporada);

        double progressoSerie = contentService.calcularProgressoSerie(conteudoId, user.getId());

        return ProgressResponse.builder()
                .episodioId(episodioId)
                .conteudoId(conteudoId)
                .status(status.name())
                .progressoTemporada(progressoTemporada)
                .progressoSerie(progressoSerie)
                .build();
    }

    @Transactional
    public ProgressResponse marcarFilme(Long conteudoId, ProgressStatus status, String userEmail) {
        User user = userService.findByEmail(userEmail);
        Content content = contentRepository.findById(conteudoId)
                .orElseThrow(() -> new EntityNotFoundException("Conteúdo não encontrado"));

        UserProgress progress = progressRepository
                .findByUsuarioIdAndConteudoId(user.getId(), conteudoId)
                .orElse(UserProgress.builder()
                        .usuario(user)
                        .conteudo(content)
                        .build());

        progress.setStatus(status);
        progress.setAtualizadoEm(LocalDateTime.now());
        progressRepository.save(progress);

        return ProgressResponse.builder()
                .conteudoId(conteudoId)
                .status(status.name())
                .progressoTemporada(status == ProgressStatus.ASSISTIDO ? 100.0 : 0.0)
                .progressoSerie(status == ProgressStatus.ASSISTIDO ? 100.0 : 0.0)
                .build();
    }
}

File: backend/src/progress/progress.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { ProgressStatus } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';
import { UserService } from '../user/user.service';
import { CatalogService } from '../catalog/catalog.service';

@Injectable()
export class ProgressService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly userService: UserService,
    private readonly catalogService: CatalogService,
  ) {}

  async marcarEpisodio(episodioId: number, status: ProgressStatus, userEmail: string) {
    const user = await this.userService.findByEmail(userEmail);
    const episode = await this.prisma.episode.findUnique({
      where: { id: episodioId },
      include: { temporada: true },
    });
    if (!episode) {
      throw new NotFoundException('Episódio não encontrado');
    }

    const existente = await this.prisma.userProgress.findUnique({
      where: { usuarioId_episodioId: { usuarioId: user.id, episodioId } },
    });

    if (existente) {
      await this.prisma.userProgress.update({
        where: { id: existente.id },
        data: { status, atualizadoEm: new Date() },
      });
    } else {
      await this.prisma.userProgress.create({
        data: { usuarioId: user.id, episodioId, status, atualizadoEm: new Date() },
      });
    }

    const conteudoId = episode.temporada.conteudoId;
    const temporadaId = episode.temporada.id;

    const totalTemporada = await this.prisma.episode.count({ where: { temporadaId } });
    const assistidosTemporada = await this.prisma.userProgress.count({
      where: { usuarioId: user.id, status: ProgressStatus.ASSISTIDO, episodio: { temporadaId } },
    });
    const progressoTemporada = totalTemporada === 0 ? 0.0 : (100.0 * assistidosTemporada) / totalTemporada;

    const progressoSerie = await this.catalogService.calcularProgressoSerie(conteudoId, user.id);

    return {
      episodioId,
      conteudoId,
      status,
      progressoTemporada,
      progressoSerie,
    };
  }

  async marcarFilme(conteudoId: number, status: ProgressStatus, userEmail: string) {
    const user = await this.userService.findByEmail(userEmail);
    const content = await this.prisma.content.findUnique({ where: { id: conteudoId } });
    if (!content) {
      throw new NotFoundException('Conteúdo não encontrado');
    }

    const existente = await this.prisma.userProgress.findFirst({
      where: { usuarioId: user.id, conteudoId },
    });

    if (existente) {
      await this.prisma.userProgress.update({
        where: { id: existente.id },
        data: { status, atualizadoEm: new Date() },
      });
    } else {
      await this.prisma.userProgress.create({
        data: { usuarioId: user.id, conteudoId, status, atualizadoEm: new Date() },
      });
    }

    const progresso = status === ProgressStatus.ASSISTIDO ? 100.0 : 0.0;
    return {
      conteudoId,
      status,
      progressoTemporada: progresso,
      progressoSerie: progresso,
    };
  }
}
